File: ClassEval_T/cpp/solution/PushBoxGame.cpp
#include <vector>
#include <string>
#include <utility>

class PushBoxGame {
public:
    PushBoxGame(const std::vector<std::string>& map);
    PushBoxGame();

    std::vector<std::string> gat_map() const;
    bool is_game_over() const;
    int get_player_col() const;
    int get_player_row() const;
    std::vector<std::pair<int, int>> get_targets() const;
    std::vector<std::pair<int, int>> get_boxes() const;
    int get_target_count() const;
    bool check_win();
    bool move(char direction);

    void init_game();

    std::vector<std::string> map;
    int player_row;
    int player_col;
    std::vector<std::pair<int, int>> targets;
    std::vector<std::pair<int, int>> boxes;
    int target_count;
    bool _is_game_over;
};


PushBoxGame::PushBoxGame(const std::vector<std::string>& map)
    : map(map), player_row(0), player_col(0), target_count(0), _is_game_over(false) {
    init_game();
}

PushBoxGame::PushBoxGame()
    : map({}), player_row(0), player_col(0), target_count(0), _is_game_over(false) {
    init_game();
}

std::vector<std::string> PushBoxGame::gat_map() const {
    return map;
}

bool PushBoxGame::is_game_over() const {
    return _is_game_over;
}

int PushBoxGame::get_player_col() const {
    return player_col;
}

int PushBoxGame::get_player_row() const {
    return player_row;
}

std::vector<std::pair<int, int>> PushBoxGame::get_targets() const {
    return targets;
}

std::vector<std::pair<int, int>> PushBoxGame::get_boxes() const {
    return boxes;
}

int PushBoxGame::get_target_count() const {
    return target_count;
}

void PushBoxGame::init_game() {
    for (size_t row = 0; row < map.size(); ++row) {
        for (size_t col = 0; col < map[row].size(); ++col) {
            if (map[row][col] == 'O') {
                player_row = static_cast<int>(row);
                player_col = static_cast<int>(col);
            }
            else if (map[row][col] == 'G') {
                targets.emplace_back(row, col);
                ++target_count;
            }
            else if (map[row][col] == 'X') {
                boxes.emplace_back(row, col);
            }
        }
    }
}

bool PushBoxGame::check_win() {
    int box_on_target_count = 0;
    for (const auto& box : boxes) {
        if (std::find(targets.begin(), targets.end(), box) != targets.end()) {
            ++box_on_target_count;
        }
    }
    if (box_on_target_count == target_count) {
        _is_game_over = true;
    }
    return _is_game_over;
}

bool PushBoxGame::move(char direction) {
    int new_player_row = player_row;
    int new_player_col = player_col;

    if (direction == 'w') {
        --new_player_row;
    }
    else if (direction == 's') {
        ++new_player_row;
    }
    else if (direction == 'a') {
        --new_player_col;
    }
    else if (direction == 'd') {
        ++new_player_col;
    }

    if (map[new_player_row][new_player_col] != '#') {
        if (std::find(boxes.begin(), boxes.end(), std::make_pair(new_player_row, new_player_col)) != boxes.end()) {
            int new_box_row = new_player_row + (new_player_row - player_row);
            int new_box_col = new_player_col + (new_player_col - player_col);

            if (map[new_box_row][new_box_col] != '#') {
                auto box_it = std::find(boxes.begin(), boxes.end(), std::make_pair(new_player_row, new_player_col));
                if (box_it != boxes.end()) {
                    boxes.erase(box_it);
                }
                boxes.emplace_back(new_box_row, new_box_col);
                player_row = new_player_row;
                player_col = new_player_col;
            }
        }
        else {
            player_row = new_player_row;
            player_col = new_player_col;
        }
    }

    return check_win();
}

File: ClassEval_T/java/solutuon/PushBoxGame.java
package org.example;

import java.util.*;

public class PushBoxGame {
    private char[][] map;
    private int playerRow;
    private int playerCol;
    private List<int[]> targets;
    private List<int[]> boxes;
    private int targetCount;
    private boolean isGameOver;

    public PushBoxGame(char[][] map) {
        this.map = new char[map.length][];
        for (int i = 0; i < map.length; i++) {
            this.map[i] = map[i].clone();
        }
        this.playerRow = 0;
        this.playerCol = 0;
        this.targets = new ArrayList<>();
        this.boxes = new ArrayList<>();
        this.targetCount = 0;
        this.isGameOver = false;
        initGame();
    }

    private void initGame() {
        for (int row = 0; row < map.length; row++) {
            for (int col = 0; col < map[row].length; col++) {
                if (map[row][col] == 'O') {
                    this.playerRow = row;
                    this.playerCol = col;
                } else if (map[row][col] == 'G') {
                    this.targets.add(new int[]{row, col});
                    this.targetCount++;
                } else if (map[row][col] == 'X') {
                    this.boxes.add(new int[]{row, col});
                }
            }
        }
    }

    public boolean checkWin() {
        int boxOnTargetCount = 0;
        for (int[] box : boxes) {
            for (int[] target : targets) {
                if (Arrays.equals(box, target)) {
                    boxOnTargetCount++;
                }
            }
        }
        isGameOver = (boxOnTargetCount == targetCount);
        return isGameOver;
    }

    public boolean move(char direction) {
        int newPlayerRow = playerRow;
        int newPlayerCol = playerCol;

        if (direction == 'w') {
            newPlayerRow -= 1;
        } else if (direction == 's') {
            newPlayerRow += 1;
        } else if (direction == 'a') {
            newPlayerCol -= 1;
        } else if (direction == 'd') {
            newPlayerCol += 1;
        }

        if (isValidMove(newPlayerRow, newPlayerCol)) {
            if (isBoxAt(newPlayerRow, newPlayerCol)) {
                int[] boxPosition = getBoxPosition(newPlayerRow, newPlayerCol);
                int newBoxRow = newPlayerRow + (newPlayerRow - playerRow);
                int newBoxCol = newPlayerCol + (newPlayerCol - playerCol);

                if (isValidMove(newBoxRow, newBoxCol) && !isBoxAt(newBoxRow, newBoxCol)) {
                    moveBox(boxPosition, newBoxRow, newBoxCol);
                    playerRow = newPlayerRow;
                    playerCol = newPlayerCol;
                }
            } else {
                playerRow = newPlayerRow;
                playerCol = newPlayerCol;
            }
        }

        return checkWin();
    }

    private boolean isValidMove(int row, int col) {
        return row >= 0 && row < map.length && col >= 0 && col < map[row].length && map[row][col] != '#';
    }

    private boolean isBoxAt(int row, int col) {
        for (int[] box : boxes) {
            if (Arrays.equals(box, new int[]{row, col})) {
                return true;
            }
        }
        return false;
    }

    private int[] getBoxPosition(int row, int col) {
        for (int[] box : boxes) {
            if (Arrays.equals(box, new int[]{row, col})) {
                return box;
            }
        }
        return null;
    }

    private void moveBox(int[] boxPosition, int newRow, int newCol) {
        for (int i = 0; i < boxes.size(); i++) {
            if (Arrays.equals(boxes.get(i), boxPosition)) {
                boxes.set(i, new int[]{newRow, newCol});
                break;
            }
        }
    }

    public boolean isGameOver() {
        return isGameOver;
    }

    public int getPlayerCol() {
        return playerCol;
    }

    public int getPlayerRow() {
        return playerRow;
    }

    public List<int[]> getTargets() {
        return new ArrayList<>(targets);
    }

    public int getTargetCount() {
        return targetCount;
    }

    public char[][] getMap() {
        return map;
    }

    public List<int[]> getBoxes() {
        return new ArrayList<>(boxes);
    }
}
